rfcs/subsuites.md

```diff
@@ -0,0 +1,348 @@
+   itself, and update the existing test scheduling logic to handle
+   multiple browser configurations.
+
+In practice people are currently using wptrunner as a monolith, with
+test selection and browser setup provided by command line options and
+configuration files, so although the architecture would probably allow
+adopting a more service-based model, it's more in keeping with existing
+usage patterns to implement support for multiple configurations
+directly in wptrunner.
+
+### Implementation Strategy
+
+From an implementation point of view, the main constraint is around
+how we associate test results with a specific configuration.
+
+Our test data flow is built around mozlog. This groups tests into
+"suites". Each suite is started with a `suite_start` event ends with a
+`suite_end` event. The `suite_start` event contains information,
+particularly in the `run_info` key about the browser being tested. That
+`run_info` data is used in wptrunner metadata files to determine which
+metadata applies to a specific test run (e.g. to allow different
+expected results on different platforms). Specific test results are
+associated to a suite implicitly by sequence i.e. any `test_start` log
+event coming after a specific `suite_start` is grouped within that
+suite. This implicitly means that only a single suite can be running
+in the context of a specific logger at a specific time. That is
+incompatible with the requirement to run multiple configurations in
+parallel.
+
+There are three obvious approaches we could use to address this:
+
+1. Use multiple loggers, one per configuration. In this case each
+   configuration would produce a separate stream of log events that
+   could be written to a separate file. However mozlog is really
+   designed around the assumption of a single top-level logger per
+   process, which is reflected in the way output is configured with
+   command line options. Since we'd presumably still want a single
+   top-level output to stdout, we'd still need to solve the problem of
+   combining multiple log streams into one top-level stream, which
+   would require a mechanism to relax the one-suite-at-a-time
+   constraint. We would also need to work out how to pass specific
+   configuration (e.g. output filenames) down to specific loggers.
+
+1. Put configuration information inside the test id, so that from the
+   point of view of consumers, the same test run in different
+   configurations look like different tests. This is how the Chromium
+   CI system handles this problem with the concept of "virtual"
+   testsuites. It has the advantage that it requires minimal changes
+   to data structures; tests get an id like `/<prefix>/<configuration
+   name>/<actual test id>`, so from the point of view of consumers
+   these are different tests. However this is also a disadvantage;
+   every consumer that wants to process the results (e.g. to compare
+   results between configurations) has to know about the details of
```

Comment:
We don't compare results between configurations in a large scale. Developers will compare results between virtual/non-virtual tests by compare the output of the two, but this is done manually.

I actually have a concern that all expectations from all subsuite to a single file could make that file too complicated. But I will have to wait to see what are the dev's opinions.

Reply:
Comparing results between configurations is definitely something we are interested in.

Note that we don't have any proposal so far that would naturally allow writing multiple log files. But with this proposal it would be relatively easy to write a log handler that reads the log messages and demuxes them to multiple files, if that's what you want.

Reply:
Thanks. For now we will keep that in a single file and see what is devs' opinion.